libs/@local/codec/src/serde/string_hash_map.rs

```diff
@@ -0,0 +1,155 @@
+//! Serialization and deserialization for a map which uses strings as keys in the serialized
+//! representation.
+//!
+//! This module provides a serializer and deserializer for [`HashMap`]s where the keys are
+//! represented as strings in the serialized form. This is useful when the keys are not strings
+//! themselves, but can be converted to and from strings.
+//!
+//! # Example
+//!
+//! ```rust
+//! use std::collections::HashMap;
+//!
+//! use serde::{Deserialize, Serialize};
+//! use serde_json::json;
+//!
+//! #[derive(Debug, PartialEq, Serialize, Deserialize)]
+//! struct MyStruct {
+//!     #[serde(with = "codec::serde::string_hash_map")]
+//!     map: HashMap<u32, String>,
+//! }
+//!
+//! let my_struct = MyStruct {
+//!     map: HashMap::from([(1, "one".to_owned()), (2, "two".to_owned())]),
+//! };
+//!
+//! let serialized = serde_json::to_value(&my_struct)?;
+//! assert_eq!(serialized, json!({"map":{ "1": "one", "2": "two" }}));
+//!
+//! let deserialized: MyStruct = serde_json::from_value(serialized)?;
+//! assert_eq!(deserialized, my_struct);
+//! # Ok::<_, serde_json::Error>(())
+//! ```
+
+use std::{
+    collections::HashMap,
+    fmt::{self, Display},
+    hash::{BuildHasher, Hash},
+    marker::PhantomData,
+    str::FromStr,
+};
+
+use serde::{
+    de::{self, Deserialize, DeserializeSeed, Deserializer, MapAccess, Visitor},
+    Serialize, Serializer,
+};
+
+use crate::serde::size_hint;
+
+/// The default serializer for [`HashMap`].
+///
+/// This is only here to allow the use of `#[serde(with = "string_hash_map")]`.
+///
+/// # Errors
+///
+/// This function can return any error that can be returned by serializing a [`HashMap`].
+#[inline]
+pub fn serialize<K, V, H, S>(map: &HashMap<K, V, H>, serializer: S) -> Result<S::Ok, S::Error>
```

Comment:
Wouldn't that lead to an asymmetrical encoding? What if the key doesn't coerce to a string? Should display or smth be used here?

Reply:
I'm not sure if we want to add the overhead of converting it to a string. The implementation of `HashMap` is already asymmetrical. It's serializing to a string as key, but does not deserialize from a string to e.g. `usize`.

Reply:
I see yea then that's fine. Makes sense